### ISSUE TITLE
Default initialization for std::random_device fails if too many connections opened at once

### ISSUE THREAD

**Issue by rmenke-lb** [scylladb/seastar#886]
If too many calls to `posix_socket_impl::find_port_and_connect` are made in a short time, `std::random_device` under older versions of libstdc++ on x86-64 will fail by throwing `std::runtime_error` with the message "random_device: rdseed failed". This is a known issue with the GNU C++ library and can be avoided by explicitly constructing `std::random_device` with the token `"rdrand"`. This token is only available on x86-64.

**Comment by nyh** [comment]
The problem is **not** making many `find_port_and_connect` calls in a short time - in `src/net/posix-stack.cc`, we only read from `std::random_device` **once** per thread, to initialize a single random engine per thread:
```c++
        static thread_local std::default_random_engine random_engine{std::random_device{}()};
        static thread_local std::uniform_int_distribution<uint16_t> u(49152/smp::count + 1, 65535/smp::count - 1);
```
So it doesn't matter if `find_port_and_connect` is called frequently - `random_device` will only be used once per thread.
But the key phrase here is "per thread" - if you have N cores, when all of them get their first request around the first time, this will cause N calls from N threads around the same time.  And if N is high enough, this can indeed cause problems if the C++ library is broken as you noticed.

This was a known bug in the C++ library: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94087
It was already fixed a year ago, and backported to gcc 10.2 at @avikivity's request. 

So this bug should not exist in libstdc++ from gcc earlier than 10 or 10.2 or newer. Are you seeing this with gcc 10.1? Can you upgrade to 10.2?

We **could** work around this bug in Seastar - write a random-number-getting function which tries random_device{}() and if it throws, retry several times or try some less secure random-number generator. But I'm not sure it's worth the effort if it can be easily fixed by updating the C++ library.

**Comment by rmenke-lb** [comment]
Understood fully. We have a workaround so it wasn't a showstopper.
